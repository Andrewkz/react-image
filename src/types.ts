export type ObserverOptions = {
	root?: Element | null;
	rootMargin?: string;
	threshold?: number;
};

export type Props = {
	alt?: string;
	options?: ObserverOptions;
	style?: Record<string, unknown>;
	placeholder?: string;
	className?: string;
	errorImage?: string;
};

export type IntersectionObserverEntryType = IntersectionObserverEntry & {
	isIntersecting: boolean;
};

export type IProps = {
	url: string;
	placeholder?: string;
	errorImage?: string;
	className?: string;
	onLoad?: () => void;
	options?: ObserverOptions;
	style?: Record<string, unknown>;
};

export type ImgProps = JSX.IntrinsicElements['img'] &
	Props & {
		height: number | string;
		width: number | string;
	};

export type DivProps = JSX.IntrinsicElements['div'] &
	Props & {
		height?: number | string;
		width?: number | string;
	};
