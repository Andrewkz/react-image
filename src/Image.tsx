import React,{useEffect, useRef, CSSProperties} from 'react'

interface info {
	src: string
	imgStyle?: CSSProperties
	alt?: string
}

export const Image: React.FC<info> = ({src, imgStyle, alt}) => {
	const imgRef = useRef<HTMLImageElement>(null)

	useEffect(() => {
		const target = imgRef.current
		if (target === null) {
			return;
		}
		if (!window.IntersectionObserver) require('intersection-observer');
		const imageObserver = new IntersectionObserver((entries) => {
			const currentTargetSrc = target.getAttribute('src');
			entries.forEach((item) => {
				if (
					(item.isIntersecting || item.intersectionRatio > 0) &&
					!currentTargetSrc
				) {
					target.setAttribute('src', src);
					target.setAttribute('alt', alt || '');
				}
			});
		});
		if (target.getAttribute('src')) {
			imageObserver.unobserve(target);
			return;
		} else {
			imageObserver.observe(target);
		}
	}, [src, alt]);

	return <img ref={imgRef} style={imgStyle} alt=""></img>;
};
