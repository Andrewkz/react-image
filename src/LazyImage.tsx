import React, { useRef, useLayoutEffect } from 'react';
import { ObserverOptions, Props } from './types';

const defaultObserverOptions: ObserverOptions = {
	root: null,
	rootMargin: '0px 0px 0px 0px',
	threshold: 0,
};

export const LazyImage: React.FC<Props> = (props) => {
	const root = useRef<HTMLDivElement>(null);

	function loadImage(imagePath: string): Promise<string> {
		return new Promise((resolve, reject) => {
			const img = new Image();
			img.src = imagePath;
			img.onload = () => resolve(imagePath);
			img.onerror = (err) => reject(err);
		});
	}

	useLayoutEffect(() => {
		if (!root.current) return;
		let matchMap = new WeakMap<NodeListOf<Element>, NodeListOf<Element>>();
		let nodeMap = new WeakMap<NodeListOf<Element>, NodeListOf<Element>>();

		const matches = root.current.querySelectorAll('img[name=lazy]');
		matchMap.set(matches, matches);
		const childrenNode = Array.prototype.slice.call(matchMap.get(matches));
		nodeMap.set(childrenNode, childrenNode);

		if (!window.IntersectionObserver) require('intersection-observer');
		const observer = new IntersectionObserver(
			(entries: IntersectionObserverEntry[]) => {
				entries
					.filter(
						(entry: IntersectionObserverEntry) =>
							(entry.target as HTMLImageElement).name === 'lazy',
					)
					.forEach((entry: IntersectionObserverEntry) => {
						let element = entry.target;
						if (entry.isIntersecting || entry.intersectionRatio > 0) {
							const src = entry.target.getAttribute('data-src') as string;
							loadImage(src)
								.catch((err) => {
									element.setAttribute('src', '');
								})
								.then((res: string) => {
									element.setAttribute('src', res);
									element.removeAttribute('name');
									element.removeAttribute('data-src');
									observer.unobserve(element);
								});
						} else {
							element.setAttribute('src', props.loading || '');
						}
					});
			},
			props.options || defaultObserverOptions,
		);
		Array.prototype.slice
			.call(nodeMap.get(childrenNode) as NodeListOf<HTMLImageElement>)
			.forEach((image: HTMLImageElement) => {
				observer.observe(image);
			});
	});

	return <div ref={root}>{props.children}</div>;
};
