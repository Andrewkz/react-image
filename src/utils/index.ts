export function loadImage(imagePath: string): Promise<string> {
	return new Promise((resolve, reject) => {
		const img = new Image();
		img.src = imagePath;
		img.onload = () => resolve(imagePath);
		img.onerror = (err) => reject(err);
	});
}

export const call = (fn: () => void): void => fn && fn();

export const isNull = <T>(obj: T | null): obj is null => obj === null;

export function setImage(element: HTMLImageElement, imagePath: string): void {
	element.src = imagePath;
	// 	element.style.backgroundImage = `url('${imagePath}')`;
}

export function addCssClassName(
	element: HTMLImageElement | HTMLDivElement,
	cssClassName: string,
): void {
	if (!element.className.includes(cssClassName)) {
		element.className += ` ${cssClassName}`;
	}
}
