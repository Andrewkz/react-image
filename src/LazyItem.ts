import { createElement, useLayoutEffect, useRef } from 'react';
import {
	ObserverOptions,
	IntersectionObserverEntryType,
	IProps,
} from './types';
import { loadImage, call, isNull, setImage } from './utils/index';

const defaultObserverOptions: ObserverOptions = {
	root: undefined,
	rootMargin: undefined,
	threshold: undefined,
};

const observerKeys: ObserverOptions[] = [];
const observers = new WeakMap<ObserverOptions, IntersectionObserver>();
const images = new WeakMap<
	HTMLImageElement,
	{ observer: IntersectionObserver; options: IProps }
>();

const LazyItem = (props) => {
	const currentRef = useRef<HTMLImageElement>(null);

	useLayoutEffect(() => {
		if (isNull(currentRef.current)) return;
		initElement(currentRef.current, props);
	});

	const createIntersectionObserver = (options: ObserverOptions) =>
		new IntersectionObserver(loadingCallback, options);

	function loadingCallback(entries: IntersectionObserverEntryType[]) {
		entries
			.filter((entry) => entry.isIntersecting || entry.intersectionRatio > 0)
			.forEach((entry) => {
				const target = entry.target as HTMLImageElement;
				const metaInfo = images.get(target);
				if (!metaInfo) {
					console.warn('Could not find meta data for src');
					return;
				}
				metaInfo.observer.unobserve(target);
				loadImage(metaInfo.options.url)
					.catch(() => {
						if (metaInfo.options.errorImage) {
							return loadImage(metaInfo.options.errorImage);
						}
						return Promise.resolve(metaInfo.options.placeholder);
					})
					.catch(() => metaInfo.options.placeholder)
					.then((imagePath: string) => {
						setImage(target as HTMLImageElement, imagePath);
						call(metaInfo.options.onLoad);
					});
			});
	}

	function initElement(element: HTMLImageElement, metadata: IProps) {
		const options = metadata.options || defaultObserverOptions;
		let observerKey = observerKeys.find(
			(oKey) =>
				oKey.root === options.root &&
				oKey.rootMargin === options.rootMargin &&
				oKey.threshold === options.threshold,
		);
		if (!observerKey) {
			observerKey = options;
			observerKeys.push(observerKey);
		}
		let observer = observers.get(observerKey);
		if (!observer) {
			observer = createIntersectionObserver(observerKey);
			observers.set(observerKey, observer);
		}
		images.set(element, {
			observer,
			options: metadata,
		});
		observer.observe(element);
	}

	return createElement(
		'img',
		{
			...props,
			ref: currentRef,
		},
		props.children,
	);
};

export default LazyItem;
