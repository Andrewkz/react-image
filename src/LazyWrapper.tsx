import React, { useRef, useLayoutEffect } from 'react';
import { ObserverOptions, Props } from './types';
import { loadImage } from './utils/index';

const defaultObserverOptions: ObserverOptions = {
	root: null,
	rootMargin: '0px 0px 0px 0px',
	threshold: 0,
};

const nodeObject = Object.create({});

const LazyWrapper: React.FC<Props> = (props) => {
	const root = useRef<HTMLDivElement>(null);
	useLayoutEffect(() => {
		if (!root.current) return;
		const nodeMap = new WeakMap<
			Record<string, unknown>,
			NodeListOf<HTMLImageElement>
		>();

		const childrenNode: NodeListOf<HTMLImageElement> = Array.prototype.slice.call(
			root.current.querySelectorAll('img[name=lazy]'),
		);

		// if (!childrenNode.length) {
		// 	return console.error('Could not find img tags in children');
		// }
		nodeMap.set(nodeObject, childrenNode);

		if (!window.IntersectionObserver) require('intersection-observer');
		const observer = new IntersectionObserver(
			(entries: IntersectionObserverEntry[]) => {
				entries
					.filter(
						(entry: IntersectionObserverEntry) =>
							entry.target.getAttribute('name') === 'lazy',
					)
					.forEach((entry: IntersectionObserverEntry) => {
						const element = entry.target;
						if (entry.isIntersecting || entry.intersectionRatio > 0) {
							const src = entry.target.getAttribute('data-src') as string;

							loadImage(src)
								.catch(() => {
									element.setAttribute('src', '');
									if (props.errorImage) {
										return loadImage(props.errorImage);
									}
									return Promise.resolve(props.placeholder);
								})
								.then((res: string) => {
									element.setAttribute('src', res);
									element.removeAttribute('name');
									element.removeAttribute('data-src');
									observer.unobserve(element);
								});
						} else {
							element.setAttribute('src', props.placeholder || '');
						}
					});
			},
			props.options || defaultObserverOptions,
		);

		Array.prototype.slice
			.call(nodeMap.get(nodeObject) as NodeListOf<HTMLImageElement>)
			.forEach((image: HTMLImageElement) => {
				observer.observe(image);
			});
	});

	return React.createElement(
		'div',
		{
			...props,
			ref: root,
		},
		props.children,
	);
};

export default LazyWrapper;
