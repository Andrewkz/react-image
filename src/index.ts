import LazyWrapper from './LazyWrapper';
import LazyItem from './LazyItem';

export { LazyWrapper, LazyItem };
