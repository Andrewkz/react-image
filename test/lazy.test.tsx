import * as React from 'react';
import { LazyWrapper } from '../src/index';
import renderer from 'react-test-renderer';
// import { shallow } from 'enzyme';

const props = {
  placeholder: ''
}

describe('first', () => {
  it('should render correctly', () => {
    const component = renderer.create(
      <LazyWrapper { ...props }>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
        <img data-src='data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=' name='lazy'></img>
      </LazyWrapper>
    );
    // console.log(component.debug());
    expect(component).toMatchSnapshot();
  });
})
