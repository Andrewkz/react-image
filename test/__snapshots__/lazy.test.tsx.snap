// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`first should render correctly 1`] = `
<div
  placeholder=""
>
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
  <img
    data-src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    name="lazy"
  />
</div>
`;
